### PR TITLE
Make `system.replicas` parallel

### LINKED ISSUE #43918: `system.replicas` table can be slow
If I select all the columns from the `system.replicas` table, and if clickhouse-keeper is located in another continent, and I have many tables, SELECT query can take multiple seconds.

It should be parallelized across tables:
https://clickhouse.com/codebrowser/ClickHouse/src/Storages/System/StorageSystemReplicas.cpp.html#154

### PR DESCRIPTION
**Pull request by evillique**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Make `system.replicas` table do parallel fetches of replicas statuses. Closes https://github.com/ClickHouse/ClickHouse/issues/43918

### CODE REVIEW

File: src/Storages/System/StorageSystemReplicas.cpp [hunk 1]
```diff
@@ -151,14 +151,32 @@ Pipe StorageSystemReplicas::read(
...
+    std::vector<StorageReplicatedMergeTree::Status> statuses(tables_size);
+
+    for (size_t i = 0; i < tables_size; ++i)
     {
-        StorageReplicatedMergeTree::Status status;
-        dynamic_cast<StorageReplicatedMergeTree &>(
-            *replicated_tables
-                [(*col_database)[i].safeGet<const String &>()]
-                [(*col_table)[i].safeGet<const String &>()]).getStatus(status, with_zk_fields);
+        thread_pool.scheduleOrThrowOnError([i, &statuses, &replicated_tables, &col_database, &col_table, &with_zk_fields]
```
Comment (kssenii) [inline src/Storages/System/StorageSystemReplicas.cpp:167]: probably this `[&, i=i]` will do the same but will make it more compact

**Review by kssenii — APPROVED** [review]

File: src/Storages/System/StorageSystemReplicas.cpp (hunk 1 above)
Comment (evillique) [inline src/Storages/System/StorageSystemReplicas.cpp:167]:
```suggestion
        thread_pool.scheduleOrThrowOnError([&, i=i]
```

File: src/Core/Settings.h [hunk 2]
```diff
@@ -155,6 +155,8 @@ static constexpr UInt64 operator""_GiB(unsigned long long value)
     \
     M(Bool, allow_experimental_parallel_reading_from_replicas, false, "If true, ClickHouse will send a SELECT query to all replicas of a table. It will work for any kind on MergeTree table.", 0) \
     \
+    M(UInt64, system_replicas_fetch_threads, 16, "The maximum number of threads to fetch data for system.replicas table.", 0) \
```
Comment (tavplubix) [inline src/Core/Settings.h:158]: Do we really need a setting for this?

**Comment by tavplubix** [comment]
Also consider using MultiRead requests in `StorageReplicatedMergeTree::getStatus`. Currently it does 3 + number_of_replicas (usually 3) = 6 subsequent synchronous requests  = 6 RTT. We can replace it with 1 MultiGetChildren and 1 MultiGet subsequent requests (2 RTT, 3 times faster).

File: src/Storages/System/StorageSystemReplicas.cpp [hunk 3]
```diff
@@ -151,14 +151,32 @@ Pipe StorageSystemReplicas::read(
...
+        thread_pool.scheduleOrThrowOnError([&, i=i]
+        {
+            dynamic_cast<StorageReplicatedMergeTree &>(
+                *replicated_tables
+                    [(*col_database)[i].safeGet<const String &>()]
+                    [(*col_table)[i].safeGet<const String &>()]).getStatus(statuses[i], with_zk_fields);
+        });
+    }
 
+    thread_pool.wait();
```
Comment (tavplubix) [inline src/Storages/System/StorageSystemReplicas.cpp:176]: The server will crash with SIGSEGV if some of the tasks throw.

**Review by tavplubix — CHANGES_REQUESTED** [review]
See the comments above

File: src/Core/Settings.h (hunk 2 above)
Comment (evillique) [reply src/Core/Settings.h:158]: We could take the number of threads from the `max_treads` (and we do), but I am not sure which value should be set in the case when `max_threads = 0`. Any suggestions would be appreciated :)

Comment (tavplubix) [reply src/Core/Settings.h:158]: max_threads = 0 usually means "use all physical cores"

File: src/Storages/System/StorageSystemReplicas.cpp [hunk 4]
```diff
@@ -151,14 +152,48 @@ Pipe StorageSystemReplicas::read(
...
+                dynamic_cast<StorageReplicatedMergeTree &>(
+                *replicated_tables
+                    [(*col_database)[i].safeGet<const String &>()]
+                    [(*col_table)[i].safeGet<const String &>()]).getStatus(statuses[i], with_zk_fields);
+            }
+            catch (...)
+            {
+                tryLogCurrentException("system.replicas", "Failed to fetch system.replicas data");
 
+                /// We capture one of the exceptions to be thrown later
```
Comment (tavplubix) [inline src/Storages/System/StorageSystemReplicas.cpp:183]: It's not a problem that tasks may throw. The problem is that if some task throws then `statuses` will be destroyed before `thread_pool`. Some thread may be still working and may try to access `statuses`.

Comment (evillique) [reply src/Storages/System/StorageSystemReplicas.cpp]: Yes, I now see the problem. Changed it a bit, so now it looks like it should be ok.

File: src/Storages/System/StorageSystemReplicas.cpp [hunk 5]
```diff
@@ -151,14 +152,39 @@ Pipe StorageSystemReplicas::read(
...
+            {
+                dynamic_cast<StorageReplicatedMergeTree &>(
+                *replicated_tables
+                    [(*col_database)[i].safeGet<const String &>()]
+                    [(*col_table)[i].safeGet<const String &>()]).getStatus(statuses[i], with_zk_fields);
+            });
+        }
+        catch (...)
+        {
+            thread_pool.wait();
```
Comment (tavplubix) [inline src/Storages/System/StorageSystemReplicas.cpp:179]: It's not necessary because `~ThreadPool` calls `wait()` anyway, but okay

**Review by tavplubix — APPROVED** [review]
LGTM now, but it would be great to optimize `StorageReplicatedMergeTree::getStatus` as well

**Comment by evillique** [comment]
> Also consider using MultiRead requests in StorageReplicatedMergeTree::getStatus. Currently it does 3 + number_of_replicas (usually 3) = 6 subsequent synchronous requests = 6 RTT. We can replace it with 1 MultiGetChildren and 1 MultiGet subsequent requests (2 RTT, 3 times faster).

From the code of `getStatus()` it looks like it will be 1 MultiGetChildren, 1 standard Get, and 1 MultiExists. I will try to optimize it.

**Comment by tavplubix** [comment]
> 1 standard Get, and 1 MultiExists

Exists can be replaced with Get and error code check

**Review by tavplubix — APPROVED** [review]

**Comment by tavplubix** [comment]
It was not necessary to merge master and restart CI

**Comment by tavplubix** [comment]
Integration tests (asan) [2/6] - https://github.com/ClickHouse/ClickHouse/issues/44928 and https://github.com/ClickHouse/ClickHouse/pull/44908
Integration tests (asan) [4/6] - https://github.com/ClickHouse/ClickHouse/issues/45160
Integration tests (release) [4/4] - https://github.com/ClickHouse/ClickHouse/pull/44908
Integration tests (tsan) [2/6] - https://github.com/ClickHouse/ClickHouse/pull/44908
Stress test (asan) - https://github.com/ClickHouse/ClickHouse/issues/44910

Stress test (ubsan) - https://github.com/ClickHouse/ClickHouse/pull/45251, also there are tons of errors like 
```
2023.01.05 07:43:05.844601 [ 1677 ] {} <Error> void DB::MergeTreeBackgroundExecutor<DB::MergeMutateRuntimeQueue>::routine(DB::TaskRuntimeDataPtr) [Queue = DB::MergeMutateRuntimeQueue]: Code: 241. DB::Exception: Memory limit (total) exceeded: would use 34.33 GiB (attempt to allocate chunk of 1146964 bytes), maximum: 34.29 GiB. OvercommitTracker decision: Memory overcommit isn't used. Waiting time or overcommit denominator are set to zero. (MEMORY_LIMIT_EXCEEDED), Stack trace (when copying this message, always include the lines below):
2023.01.05 07:43:05.855458 [ 1686 ] {} <Error> void DB::MergeTreeBackgroundExecutor<DB::MergeMutateRuntimeQueue>::routine(DB::TaskRuntimeDataPtr) [Queue = DB::MergeMutateRuntimeQueue]: Code: 241. DB::Exception: Memory limit (total) exceeded: would use 34.33 GiB (attempt to allocate chunk of 1051048 bytes), maximum: 34.29 GiB. OvercommitTracker decision: Memory overcommit isn't used. Waiting time or overcommit denominator are set to zero. (MEMORY_LIMIT_EXCEEDED), Stack trace (when copying this message, always include the lines below):
2023.01.05 07:43:05.860716 [ 1681 ] {} <Error> void DB::MergeTreeBackgroundExecutor<DB::MergeMutateRuntimeQueue>::routine(DB::TaskRuntimeDataPtr) [Queue = DB::MergeMutateRuntimeQueue]: Code: 241. DB::Exception: Memory limit (total) exceeded: would use 34.33 GiB (attempt to allocate chunk of 1072930 bytes), maximum: 34.29 GiB. OvercommitTracker decision: Memory overcommit isn't used. Waiting time or overcommit denominator are set to zero. (MEMORY_LIMIT_EXCEEDED), Stack trace (when copying this message, always include the lines below):
2023.01.05 07:43:05.871276 [ 1688 ] {} <Error> void DB::MergeTreeBackgroundExecutor<DB::MergeMutateRuntimeQueue>::routine(DB::TaskRuntimeDataPtr) [Queue = DB::MergeMutateRuntimeQueue]: Code: 241. DB::Exception: Memory limit (total) exceeded: would use 34.33 GiB (attempt to allocate chunk of 1051048 bytes), maximum: 34.29 GiB. OvercommitTracker decision: Memory overcommit isn't used. Waiting time or overcommit denominator are set to zero. (MEMORY_LIMIT_EXCEEDED), Stack trace (when copying this message, always include the lines below):
2023.01.05 07:43:05.876756 [ 1683 ] {} <Error> void DB::MergeTreeBackgroundExecutor<DB::MergeMutateRuntimeQueue>::routine(DB::TaskRuntimeDataPtr) [Queue = DB::MergeMutateRuntimeQueue]: Code: 241. DB::Exception: Memory limit (total) exceeded: would use 34.33 GiB (attempt to allocate chunk of 1072930 bytes), maximum: 34.29 GiB. OvercommitTracker decision: Memory overcommit isn't used. Waiting time or overcommit denominator are set to zero. (MEMORY_LIMIT_EXCEEDED), Stack trace (when copying this message, always include the lines below):
2023.01.05 07:43:05.887665 [ 1674 ] {} <Error> void DB::MergeTreeBackgroundExecutor<DB::MergeMutateRuntimeQueue>::routine(DB::TaskRuntimeDataPtr) [Queue = DB::MergeMutateRuntimeQueue]: Code: 241. DB::Exception: Memory limit (total) exceeded: would use 34.33 GiB (attempt to allocate chunk of 1051048 bytes), maximum: 34.29 GiB. OvercommitTracker decision: Memory overcommit isn't used. Waiting time or overcommit denominator are set to zero. (MEMORY_LIMIT_EXCEEDED), Stack trace (when copying this message, always include the lines below):
2023.01.05 07:43:05.893049 [ 1677 ] {} <Error> void DB::MergeTreeBackgroundExecutor<DB::MergeMutateRuntimeQueue>::routine(DB::TaskRuntimeDataPtr) [Queue = DB::MergeMutateRuntimeQueue]: Code: 241. DB::Exception: Memory limit (total) exceeded: would use 34.33 GiB (attempt to allocate chunk of 1072930 bytes), maximum: 34.29 GiB. OvercommitTracker decision: Memory overcommit isn't used. Waiting time or overcommit denominator are set to zero. (MEMORY_LIMIT_EXCEEDED), Stack trace (when copying this message, always include the lines below):
2023.01.05 07:43:05.906151 [ 1686 ] {} <Error> void DB::MergeTreeBackgroundExecutor<DB::MergeMutateRuntimeQueue>::routine(DB::TaskRuntimeDataPtr) [Queue = DB::MergeMutateRuntimeQueue]: Code: 241. DB::Exception: Memory limit (total) exceeded: would use 34.33 GiB (attempt to allocate chunk of 1051048 bytes), maximum: 34.29 GiB. OvercommitTracker decision: Memory overcommit isn't used. Waiting time or overcommit denominator are set to zero. (MEMORY_LIMIT_EXCEEDED), Stack trace (when copying this message, always include the lines below):
2023.01.05 07:43:05.911196 [ 1681 ] {} <Error> void DB::MergeTreeBackgroundExecutor<DB::MergeMutateRuntimeQueue>::routine(DB::TaskRuntimeDataPtr) [Queue = DB::MergeMutateRuntimeQueue]: Code: 241. DB::Exception: Memory limit (total) exceeded: would use 34.33 GiB (attempt to allocate chunk of 1072930 bytes), maximum: 34.29 GiB. OvercommitTracker decision: Memory overcommit isn't used. Waiting time or overcommit denominator are set to zero. (MEMORY_LIMIT_EXCEEDED), Stack trace (when copying this message, always include the lines below):
```
so I guess it's the "deadlock" in OvercommitTracker (https://github.com/ClickHouse/ClickHouse/pull/45236)

**Pull request merged**